README
NAME
    Cookieville - REST API for your database

VERSION
    0.05

DESCRIPTION
    Cookieville is a a web application which allow you to inspect and run
    queries on your SQL database using a REST interface.

    This distribution also contain Cookieville::Client for easy blocking and
    non-blocking integration with the server.

    This application need a DBIx::Class based schema to work. It will query
    the result files on disk to gather meta information about the schema
    instead of looking into the running database.

    THIS SERVER IS CURRENTLY EXPERIMENTAL AND WILL CHANGE WITHOUT ANY
    NOTICE.

SYNOPSIS
      $ MOJO_CONFIG=/path/to/mojo.conf cookieville daemon --listen http://*:5000

    Example "MOJO_CONFIG":

      $ cat /path/to/mojo.conf
      {
        # config for cookieville
        inactive_timeout => 10,
        schema_class => "My::Schema",
        connect_args => {
          "DBI:mysql:database=some_database;host=localhost",
          "dr_who",
          "MostS3cretpassWord",
        },

        # config for hypnotoad - https://metacpan.org/pod/Mojo::Server::Hypnotoad
        hypnotoad => {
          listen => [ "http://*:5000" ],
          workers => 10,
        },

        # will set up logging to a given file
        log => {
          path => "/path/to/cookieville.log",
          level => "info", # or debug
        },
      }

    TIP: Give config file the file mode 0600 to protect your connect
    passwords.

    TIP: Run Cookieville with hypnotoad and enough workers in production.
    Reason for this is that DBIx::Class is blocking, and therefor can only
    handle one database opartion pr. worker. You might also want to tweak
    "inactive_timeout" to prevent a worker from running a query for too
    long.

RESOURCES
    *   GET /

        Returns a description of this application:

          {
            "version": "0.01",
            "source": "https://github.com/jhthorsen/cookieville",
            "resources": {
              "schema_source_list": [ "GET", "/sources" ],
              "schema_for_source": [ "GET", "/:source/schema" ],
              "source_search": [ "GET", "/:source/search?q=:json&limit=:int&order_by:json" ],
              "source_delete": [ "DELETE", "/:source/:id" ],
              "source_patch": [ "PATCH", "/:source/:id" ],
              "source_update_or_insert": [ "PUT", "/:source" ]
            }
          }

        The environment variable "COOKIEVILLE_INFO" can be used to limit the
        data returned:

          COOKIEVILLE_INFO=source,resources

    *   GET /sources

        Returns a list of available sources (resultsets). Example:

          [ "Users", "Posts" ]

    *   GET /:source/schema

        Returns the schema for the given "source".

    *   GET /:source/search

        Does a SELECT from the given "source" with a given set of query
        params:

        *   q=:json (mandatory)

            "q" will be deserialized and used as the query part.

        *   columns=:json (optional)

            Only output the given columns. Example:

              columns=["id","name"]

        *   limit=:int (optional)

            Used to limit the number of rows in the output.

        *   page=:int (optional)

            Used for pagination when "limit" is specified.

        *   order_by=:json (optional)

            Sort the result by column(s). Examples:

              order_by={"-desc","name"}
              order_by=["name","id"]

        The return value will be a JSON document containing the rows.
        Example:

          {
            data: [
              { "id": 1002, "name": "Jan Henning Thorsen", "age": 31 },
              { "id": 3005, "name": "Billy West", "age": 62 }
            ]
          }

        TODO: Make sure integers from the database are actual integers in
        the result JSON.

        The format .csv
        <http://en.wikipedia.org/wiki/Comma-separated_values> is also
        supported. Example:

          GET /Users.csv?q={"age":31}&order_by=name

    *   DELETE /:source/:id

        Used to DELETE a single record identified by "id".

        The return value will be a JSON document with the number of rows
        deleted:

          {"n":1}

        NOTE: This will be "{"n":0}" if the record was already deleted.

    *   PATCH /:source/:id

        Used to do a (partial) UPDATE of a single row identified by "id".
        The HTTP body must be a JSON structure with the data to update to.

        The return value will have the new document. Example:

          {
            "data": { "id": 1002, "name": "Jan Henning Thorsen", "age": 31 }
          }

        Will return 404 if the given "id" does not match any records in the
        database.

    *   PUT /:source

        Used to INSERT or UPDATE a single row. The HTTP body must be a JSON
        structure with the data to insert or update.

        The return value will be a JSON document containing all the data for
        the inserted or updated row. Example:

          {
            "inserted": true, # or false
            "data": { "id": 1002, "name": "Jan Henning Thorsen", "age": 31 }
          }

  Error handling
    The API will return "200 OK" on success and another error code on
    failure:

    *   400

        Return the document below on invalid input data. "message" holds a
        description of what is missing. Example:

          { "message": "Missing (q) query param." }

    *   401

        Return the document below on when not authorized. "message" holds a
        description of why not. Example:

          { "message": "Invalid token." }

    *   404

        Return the document below if the given resource could not be found.
        "message" holds a description of what is not found. Examples:

          { "message": "Resource not found" }
          { "message": "No source by that name." }
          { "message": "No matching records in database." }

    *   500

          { "message": "Internal server error." }

        Generic error when something awful happens. "message" might not make
        any sense. Look at the server log for more details.

    Other error codes might be added in future releases.

  Queries
    The queries (referred to as the "q" query param in the API) are passed
    on as the first argument to "search" in DBIx::Class.

ATTRIBUTES
  inactive_timeout
      $int = $self->inactive_timeout;

    Used to set the number of seconds before a query agains the database
    time out. Defaults to value from config, the environment variable
    "COOKIEVILLE_INACTIVE_TIMEOUT" or 10 seconds.

  connect_args
      $array_ref = $self->connect_args;

    Looks in config to find connection arguments for "schema_class". See
    "SYNOPSIS" for details.

  schema_class
      $class_name = $self->schema_class;

    Looks in config to find the schema class to use. See "SYNOPSIS" for
    details.

HELPERS
  db
      $obj = $self->db;

    Returns an instance of "schema_class".

METHODS
  setup_routes
    Used to setup the "RESOURCES".

  startup
    Will set up "RESOURCES" and add "HELPERS".

COPYRIGHT AND LICENSE
    Copyright (C) 2014, Jan Henning Thorsen

    This program is free software, you can redistribute it and/or modify it
    under the terms of the Artistic License version 2.0.

AUTHOR
    Jan Henning Thorsen - "jhthorsen@cpan.org"



Changes
Revision history for Cookieville

0.05     Fri Jun 20 15:07:48 2014
       * Add Cookieville::Plugin::AccessLog
       * Add Cookieville::Plugin::Authorize
       * Add support for "log" in MOJO_CONFIG

0.04     Fri Jun 20 09:24:38 2014
       * Add Cookieville::Client to the main Cookieville documentation
       * Remove support for reading $HOME/.cookieville config
       * Remove File::HomeDir dependency

0.03     Fri Jun 20 09:08:11 2014
       * Add Cookieville::Client
       * Fix PUT /:source on UPDATE
       * Change PUT /:source will INSERT when no unique constraints are found
       * Change "inserted" in PUT /:source result to an integer: 0 or 1.

0.02     Tue Jun 17 22:24:55 2014
       * Add support for reading MOJO_CONFIG
       * Add support for "columns" in GET /:source/search?q=:json&columns=["foo"]
       * Add support for custom inactivity timeout

0.01     Tue Jun 17 20:22:33 2014
       * EXPERIMENTAL release
       * Add resource for GET /
       * Add resource for GET /sources
       * Add resource for GET /:source/schema
       * Add resource for GET /:source/search?q=:json&limit=:int&page=:int&order_by=:json
       * Add resource for DELETE /:source/:id
       * Add resource for PATCH /:source/:id
       * Add resource for PUT /:source


t/Helper.pm
package t::Helper;
use Mojo::Base -strict;
use Mojo::Util;
use Cookieville;
use t::Schema;

my $DB_FILE;

sub client {
  my ($class, $t) = @_;
  require Cookieville::Client;
  my $client = Cookieville::Client->new(url => '/');
  $client->_ua($t->ua);
  $client;
}

sub client_cb {
  ($::err, $::res) = ('', undef);

  return sub {
    my $client = shift;
    ($::err, $::res) = @_;
    $client->_ua->ioloop->stop;
  };
}

sub t {
  my $app = Cookieville->new;
  my $t = Test::Mojo->new($app);

  $app->connect_args([ "dbi:SQLite:dbname=$DB_FILE" ]);
  $app->schema_class('t::Schema');
  $app->db->deploy;

  return $t;
}

sub import {
  my $class = shift;
  my $caller = caller;

  $DB_FILE = $0;
  $DB_FILE =~ s!^\W+!!;
  $DB_FILE =~ s!\W!_!g;
  $DB_FILE .= '.sqlite';
  unlink $DB_FILE;

  strict->import;
  warnings->import;

  Mojo::Util::monkey_patch($caller => d => \&Mojo::Util::dumper);

  eval <<"  CODE" or die $@;
    package $caller;
    use Test::Mojo;
    use Test::More;
    1;
  CODE
}

END {
  unlink $DB_FILE if $DB_FILE;
}

1;


t/Schema.pm
package t::Schema;
use base qw/DBIx::Class::Schema/;
__PACKAGE__->load_namespaces();
1;


t/Schema/Result/CD.pm
package t::Schema::Result::CD;
use base 'DBIx::Class::Core';
__PACKAGE__->table('cd');
__PACKAGE__->add_columns(
  id => { data_type => "integer", extra => { unsigned => 1 }, is_auto_increment => 1, is_nullable => 0 },
  artist_id => { data_type => "integer", extra => { unsigned => 1 }, is_nullable => 0 },
  name => { data_type => "varchar", default_value => "", is_nullable => 0, size => 80 },
);
__PACKAGE__->set_primary_key('id');
__PACKAGE__->belongs_to(artist => 't::Schema::Result::Artist', 'id');
1;


Makefile.PL
use ExtUtils::MakeMaker;
WriteMakefile(
  NAME => 'Cookieville',
  ABSTRACT_FROM => 'lib/Cookieville.pm',
  VERSION_FROM => 'lib/Cookieville.pm',
  AUTHOR => 'Jan Henning Thorsen <jhthorsen@cpan.org>',
  LICENSE => 'perl',
  PREREQ_PM => {
    'DBIx::Class' => '0.08',
    'Mojolicious' => '4.03',
    'Time::HiRes' => '1.90',
  },
  BUILD_REQUIRES => {
    'DBD::SQLite' => '1.00',
    'SQL::Translator' => '0.11016',
    'Test::More' => '0.90',
  },
  META_MERGE => {
    resources => {
      license => 'http://dev.perl.org/licenses/',
      homepage => 'https://metacpan.org/release/Cookieville',
      bugtracker => 'https://github.com/jhthorsen/cookieville/issues',
      repository => 'https://github.com/jhthorsen/cookieville.git',
    },
  },
  test => {
    TESTS => 't/*.t',
  },
  EXE_FILES => ['script/cookieville'],
);


lib/Cookieville/Read.pm
package Cookieville::Read;

=head1 NAME

Cookieville::Read - Controller for getting data from the schema

=cut

use Mojo::Base 'Mojolicious::Controller';
use Mojo::JSON 'j';

=head1 METHODS

=head2 search

Get data from a given source.

=cut

sub search {
  my $self = shift;
  my $rs = eval { $self->db->resultset($self->stash('source')) };
  my $q = j $self->param('q');
  my %extra;

  unless ($rs) {
    return $self->respond_to(
      csv => sub { shift->render(text => 'No source by that name.', status => 404); },
      txt => sub { shift->render(text => 'No source by that name.', status => 404); },
      any => sub { shift->render(json => { message => 'No source by that name.' }, status => 404); },
    );
  }
  unless (ref $q eq 'HASH') {
    return $self->respond_to(
      csv => sub { shift->render(text => 'Invalid (q) query param.', status => 400); },
      any => sub { shift->render(json => { message => 'Invalid (q) query param.' }, status => 400); },
    );
  }

  $extra{order_by} = j $self->param('order_by') if $self->param('order_by');
  $extra{page} = $self->param('page') if $self->param('page');
  $extra{rows} = $self->param('limit') if $self->param('limit');
  $extra{columns} = j $self->param('columns') if $self->param('columns');

  $self->stash(extra => \%extra, rs => $rs->search_rs($q, \%extra));
  $self->respond_to(
    csv => \&_search_as_csv,
    txt => \&_search_as_csv,
    any => \&_search_as_json,
  );
}

sub _search_as_csv {
  my $self = shift;
  my $rs = $self->stash('rs');
  my @columns;
  my @csv;

  if (my $columns = $self->stash('extra')->{columns}) {
    @columns = @$columns;
  }
  else {
    @columns = $rs->result_source->columns;
  }

  push @csv, join ',', @columns;

  while(my $row = $rs->next) {
    push @csv, join ',', map {
      my $v = $row->get_column($_);
      $v =~ s!"!""!g;
      $v =~ /[\s,]/ ? qq("$v") : $v;
    } @columns;
  }

  $self->render(text => join '', map { "$_\n" } @csv);
}

sub _search_as_json {
  my $self = shift;
  my $rs = $self->stash('rs');
  my @data;

  while(my $row = $rs->next) {
    push @data, { $row->get_columns };
  }

  $self->render(json => { data => \@data });
}

=head1 AUTHOR

Jan Henning Thorsen - C<jhthorsen@cpan.org>

=cut

1;


lib/Cookieville/Write.pm
package Cookieville::Write;

=head1 NAME

Cookieville::Write - Controller for changing data in a result

=cut

use Mojo::Base 'Mojolicious::Controller';

=head1 METHODS

=head2 delete

Delete data from the database.

=cut

sub delete {
  my $self = shift;
  my $rs = eval { $self->db->resultset($self->stash('source')) };
  my $row;

  unless ($rs) {
    return $self->render(json => { message => 'No source by that name.' }, status => 404);
  }

  $row = $rs->find($self->stash('id'));
  $row->delete if $row;

  $self->render(json => { n => $row ? 1 : 0 });
}

=head2 patch

Used to patch a record in the database.

=cut

sub patch {
  my $self = shift;
  my $source = $self->stash('source');
  my $rs = eval { $self->db->resultset($source) };
  my $data = $self->req->json;
  my $row;

  unless ($data) {
    return $self->render(json => { message => 'Invalid JSON body.' }, status => 400);
  }
  unless ($rs) {
    return $self->render(json => { message => 'No source by that name.' }, status => 404);
  }

  $row = $rs->find($self->stash('id'));

  unless ($row) {
    return $self->render(json => { message => qq(No such record in $source source.) }, status => 404);
  }

  $row->set_column($_ => $data->{$_}) for keys %$data;
  $row->update;
  $self->render(json => { data => { $row->get_columns } });
}

=head2 update_or_insert

Used to update or insert a record in the database.

=cut

sub update_or_insert {
  my $self = shift;
  my $source = $self->stash('source');
  my $rs = eval { $self->db->resultset($source) };
  my $data = $self->req->json;
  my ($in_storage, $row);

  unless ($data) {
    return $self->render(json => { message => 'Invalid JSON body.' }, status => 400);
  }
  unless ($rs) {
    return $self->render(json => { message => 'No source by that name.' }, status => 404);
  }

  if ($self->_can_find($data, $rs->result_source)) {
    $row = $rs->find_or_new($data);
    $in_storage = $row->in_storage;
    $row->in_storage ? $row->update($data) : $row->insert;
  }
  else {
    $row = $rs->create($data);
    $row->discard_changes;
    $in_storage = 0;
  }

  $self->render(
    json => {
      inserted => $in_storage ? 0 : 1,
      data => { $row->get_columns },
    },
  );
}

sub _can_find {
  my ($self, $data, $source) = @_;
  my %unique_constraints = $source->unique_constraints;

  for my $c (values %unique_constraints) {
    return 1 if @$c == grep { exists $data->{$_} } @$c;
  }

  return 0;
}

=head1 AUTHOR

Jan Henning Thorsen - C<jhthorsen@cpan.org>

=cut

1;


lib/Cookieville/Client.pm
package Cookieville::Client;

=head1 NAME

Cookieville::Client - Client that talks with Cookieville server

=head1 DESCRIPTION

L<Cookieville::Client> is a client that talks with the L<Cookieville> server.

=head1 SYNOPSIS

  use Cookieville::Client;
  my $ua = Cookieville::Client->new(url => 'http://127.0.0.1/');
  my $res;

  $res = $ua->search(
            'SomeSource',
            { col_a => { like => '%42' } },
            { limit => 10 },
          );

  $res = $ua->delete(SomeSource => 123);
  $res = $ua->put(SomeSource => { col_a => 123 });
  $res = $ua->patch(SomeSource => 42 => { col_a => 123 });

=head2 Error handling

Blocking requests will throw an exception on error, while all callbacks
receive the error as a string. Example:

  use Mojolicious::Lite;

  get '/artists' => sub {
    my $c = shift->render_later;

    Mojo::IOLoop->delay(
      sub {
        my ($delay) = @_;
        $c->cookieville_helper->search(Artist => {}, $delay->begin);
      },
      sub {
        my ($delay, $err, $res) = @_;
        return $c->render_exception($err) if $err;
        return $c->render(json => $res->{data});
      },
    );
  };

=cut

use Mojo::Base -base;
use Mojo::JSON 'j';
use Mojo::UserAgent;
use Mojo::URL;
use Carp ();

=head1 ATTRIBUTES

=head2 id

  $str = $self->id;
  $self = $self->id($str);

Used to set the "X-Cookieville-Auth-Id" HTTP header. This can be used by the
L<Cookieville::Plugin::Authorize> plugin.

Defaults to C<COOKIEVILLE_AUTH_ID> environment variable.

=head2 url

  $url_obj = $self->url;
  $self = $self->url($url_obj);

Holds the base URL to the L<Cookieville> server.
Default to "http://127.0.0.1/".

=cut

has id => $ENV{COOKIEVILLE_AUTH_ID} || '';
has url => sub { Mojo::URL->new('http://127.0.0.1/'); };
has _ua => sub { Mojo::UserAgent->new; };

=head1 METHODS

=head2 new

Will make sure "url" in constructor is indeed a L<Mojo::URL> object.

=cut

sub new {
  my $self = shift->SUPER::new(@_);

  $self->url(Mojo::URL->new($self->url)) if $self->url;
  $self;
}

=head2 delete

  $res = $self->delete($source => $id);
  $self = $self->delete($source => $id, $cb);

Used to DELETE a single record from C<$source>, identified by C<id>.

=cut

sub delete {
  my ($self, $source, $id, $cb) = @_;

  return $self->_blocking(delete => $source, $id) unless ref $cb eq 'CODE';
  return $self->_abort('Invalid source or id.', $cb) unless $source and defined $id;

  Scalar::Util::weaken($self);
  $self->_ua->delete(
    $self->_url("/$source/$id"),
    { 'X-Cookieville-Auth-Id' => $self->id },
    sub { $self->$cb($self->_res_from_tx($_[1])); },
  );

  return $self;
}

=head2 patch

  $res = $self->patch($source => $id => \%data);
  $self = $self->patch($source => $id => \%data, $cb);

Used to UPDATE a single record from C<$source>, identified by C<id>.
C<%data> can be partial or full set of column/values.

=cut

sub patch {
  my ($self, $source, $id, $data, $cb) = @_;

  return $self->_blocking(patch => $source, $id, $data) unless ref $cb eq 'CODE';
  return $self->_abort('Invalid source or id.', $cb) unless $source and defined $id;
  return $self->_abort('Invalid data.', $cb) unless ref $data eq 'HASH';

  Scalar::Util::weaken($self);
  $self->_ua->patch(
    $self->_url("/$source/$id"),
    { 'X-Cookieville-Auth-Id' => $self->id },
    j($data),
    sub { $self->$cb($self->_res_from_tx($_[1])); },
  );

  return $self;
}

=head2 put

  $res = $self->put($source => \%data);
  $self = $self->put($source => \%data, $cb);

Used to INSERT or UPDATE a single row. An UPDATE will be issued if C<%data>
contain an unique constraint a matching record in database.

L</put> v.s L</patch>: L</patch> will never INSERT a new record, while
L</put> will make sure a given record exists.

NOTE: C<%data> without any unique constraints will result in INSERT.

=cut

sub put {
  my ($self, $source, $data, $cb) = @_;

  return $self->_blocking(put => $source, $data) unless ref $cb eq 'CODE';
  return $self->_abort('Invalid source.', $cb) unless $source;
  return $self->_abort('Invalid data.', $cb) unless ref $data eq 'HASH';

  Scalar::Util::weaken($self);
  $self->_ua->put(
    $self->_url("/$source"),
    { 'X-Cookieville-Auth-Id' => $self->id },
    j($data),
    sub { $self->$cb($self->_res_from_tx($_[1])); },
  );

  return $self;
}

=head2 search

  $res = $self->search($source => \%query, \%extra);
  $self = $self->search($source => \%query, \%extra, $cb);

Does a SELECT from the given C<source> with a given C<%query> and C<%extra>
parameters. This method is very similar to L<DBIx::Class::ResultSet/search>,
but with less C<%extra> options:

=over 4

=item * columns

Only output the given columns. Example:

  $extra{columns} = [qw( id name )];

=item * limit

Used to limit the number of rows in the output.

  $extra{limit} = 10;

=item * page=:int (optional)

Used for pagination when C<limit> is specified.

  $extra{limit} = 2;

=item * order_by

Sort the result by column(s). Examples:

  $extra{order_by} = ["name"];
  $extra{order_by} = { "-desc" => "name" };

=back

=cut

sub search {
  my $cb = ref $_[-1] eq 'CODE' ? pop : undef;
  my $self = shift;
  my $source = shift;
  my $query = shift;
  my $extra = shift || {};

  return $self->_blocking(search => $source, $query, $extra) unless ref $cb eq 'CODE';
  return $self->_abort('Invalid source.', $cb) unless $source;
  return $self->_abort('Invalid query.', $cb) unless ref($query) =~ /HASH|ARRAY/;

  $extra->{columns} = j $extra->{columns} if $extra->{columns};
  $extra->{order_by} = j $extra->{order_by} if $extra->{order_by};
  $extra->{q} = j $query;

  Scalar::Util::weaken($self);
  $self->_ua->get(
    $self->_url("/$source/search")->query($extra),
    { 'X-Cookieville-Auth-Id' => $self->id },
    sub { $self->$cb($self->_res_from_tx($_[1])); },
  );

  return $self;
}

sub _blocking {
  my ($self, $method, @args) = @_;
  my ($err, $res);

  $self->$method(@args, sub {
    my $self = shift;
    ($err, $res) = @_;
    $self->_ua->ioloop->stop;
  });

  $self->_ua->ioloop->start;
  Carp::confess($err) if $err;
  return $res;
}

sub _abort {
  my ($self, $err, $cb) = @_;

  Scalar::Util::weaken($self);
  $self->_ua->ioloop->timer(0, sub { $self and $self->$cb($err) });
  $self;
}

sub _res_from_tx {
  my ($self, $tx) = @_;
  my ($err, $res);

  $res = $tx->res->json;

  if (not ref $res) {
    $err = $tx->error;
    $err = $err->{message} if ref $err;
  }
  elsif ($tx->res->code != 200) {
    $err = $res->{message};
  }

  return $err, $res || {};
}

sub _url {
  $_[0]->url->clone->path($_[1]);
}

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014, Jan Henning Thorsen

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

=head1 AUTHOR

Jan Henning Thorsen - C<jhthorsen@cpan.org>

=cut

1;


lib/Cookieville/Schema.pm
package Cookieville::Schema;

=head1 NAME

Cookieville::Schema - Controller for running actions on the schema

=cut

use Mojo::Base 'Mojolicious::Controller';

=head1 METHODS

=head2 index

Render information about this application.

=cut

sub index {
  my $self = shift;

  $self->render(
    json => {
      map { $_ => $self->${ \ "_info_$_" } } split /,/, +($ENV{COOKIEVILLE_INFO} // 'version,source,resources')
    },
  );
}

=head2 sources_list

Render a list of sources.

=cut

sub sources_list {
  my $self = shift;
  my @sources = sort $self->db->sources;

  $self->respond_to(
    csv => sub { shift->render(text => join ",", @sources); },
    txt => sub { shift->render(text => join "", map { "$_\n" } @sources); },
    any => sub { shift->render(json => \@sources); },
  );
}

=head2 source_schema

Render the source definition.

=cut

sub source_schema {
  my $self = shift;
  my $source = eval { $self->db->source($self->stash('source')) };

  if ($source) {
    $self->render(
      json => {
        columns => $self->_columns_from($source),
        name => $source->name,
        primary_columns => [ $source->primary_columns ],
      },
    );
  }
  else {
    $self->render(json => { message => 'No source by that name.' }, status => 404);
  }
}

sub _columns_from {
  my ($self, $source) = @_;
  my %columns;

  for my $name ($source->columns) {
    $columns{$name} = $source->column_info($name);
  }

  return \%columns;
}

sub _info_resources {
  +{
    schema_source_list => [ "GET", "/sources" ],
    schema_for_source => [ "GET", "/:source/schema" ],
    source_select => [ "GET", "/:source/select?q=:json&limit=:int&order_by:json" ],
    source_delete => [ "DELETE", "/:source/:id" ],
    source_patch => [ "PATCH", "/:source/:id" ],
    source_update_or_insert => [ "PUT", "/:source" ],
  };
}

sub _info_source {
  'https://github.com/jhthorsen/cookieville';
}

sub _info_version {
  eval { Cookieville->VERSION } || 0;
}

=head1 AUTHOR

Jan Henning Thorsen - C<jhthorsen@cpan.org>

=cut

1;


lib/Cookieville.pm
package Cookieville;

=head1 NAME

Cookieville - REST API for your database

=head1 VERSION

0.05

=head1 DESCRIPTION

L<Cookieville> is a a web application which allow you to inspect and run
queries on your SQL database using a REST interface.

This distribution also contain L<Cookieville::Client> for easy blocking and
non-blocking integration with the server.

This application need a L<DBIx::Class> based L<schema|/schema_class> to work.
It will query the result files on disk to gather meta information about the
schema instead of looking into the running database.

THIS SERVER IS CURRENTLY EXPERIMENTAL AND WILL CHANGE WITHOUT ANY NOTICE.

=head1 SYNOPSIS

  $ MOJO_CONFIG=/path/to/mojo.conf cookieville daemon --listen http://*:5000

Example C<MOJO_CONFIG>:

  $ cat /path/to/mojo.conf
  {
    # config for cookieville
    inactive_timeout => 10,
    schema_class => "My::Schema",
    connect_args => {
      "DBI:mysql:database=some_database;host=localhost",
      "dr_who",
      "MostS3cretpassWord",
    },

    # config for hypnotoad - https://metacpan.org/pod/Mojo::Server::Hypnotoad
    hypnotoad => {
      listen => [ "http://*:5000" ],
      workers => 10,
    },

    # will set up logging to a given file
    log => {
      path => "/path/to/cookieville.log",
      level => "info", # or debug
    },
  }

TIP: Give config file the file mode 0600 to protect your connect passwords.

TIP: Run L<Cookieville> with L<hypnotoad|Mojo::Server::Hypnotoad> and enough
workers in production. Reason for this is that L<DBIx::Class> is blocking,
and therefor can only handle one database opartion pr. worker. You might
also want to tweak L</inactive_timeout> to prevent a worker from running
a query for too long.

=head1 RESOURCES

=over 4

=item * GET /

Returns a description of this application:

  {
    "version": "0.01",
    "source": "https://github.com/jhthorsen/cookieville",
    "resources": {
      "schema_source_list": [ "GET", "/sources" ],
      "schema_for_source": [ "GET", "/:source/schema" ],
      "source_search": [ "GET", "/:source/search?q=:json&limit=:int&order_by:json" ],
      "source_delete": [ "DELETE", "/:source/:id" ],
      "source_patch": [ "PATCH", "/:source/:id" ],
      "source_update_or_insert": [ "PUT", "/:source" ]
    }
  }

The environment variable C<COOKIEVILLE_INFO> can be used to limit the data returned:

  COOKIEVILLE_INFO=source,resources

=item * GET /sources

Returns a list of available sources (resultsets). Example:

  [ "Users", "Posts" ]

=item * GET /:source/schema

Returns the schema for the given C<source>.

=item * GET /:source/search

Does a SELECT from the given C<source> with a given set of query params:

=over 4

=item * q=:json (mandatory)

C<q> will be L<deserialized|Mojo::JSON/decode> and used as the
L<query part|/Queries>.

=item * columns=:json (optional)

Only output the given columns. Example:

  columns=["id","name"]

=item * limit=:int (optional)

Used to limit the number of rows in the output.

=item * page=:int (optional)

Used for pagination when C<limit> is specified.

=item * order_by=:json (optional)

Sort the result by column(s). Examples:

  order_by={"-desc","name"}
  order_by=["name","id"]

=back

The return value will be a JSON document containing the rows. Example:

  {
    data: [
      { "id": 1002, "name": "Jan Henning Thorsen", "age": 31 },
      { "id": 3005, "name": "Billy West", "age": 62 }
    ]
  }

TODO: Make sure integers from the database are actual integers in the
result JSON.

The format L<.csv|http://en.wikipedia.org/wiki/Comma-separated_values> is
also supported. Example:

  GET /Users.csv?q={"age":31}&order_by=name

=item * DELETE /:source/:id

Used to DELETE a single record identified by C<id>.

The return value will be a JSON document with the number of rows deleted:

  {"n":1}

NOTE: This will be C<{"n":0}> if the record was already deleted.

=item * PATCH /:source/:id

Used to do a (partial) UPDATE of a single row identified by C<id>. The HTTP
body must be a JSON structure with the data to update to.

The return value will have the new document. Example:

  {
    "data": { "id": 1002, "name": "Jan Henning Thorsen", "age": 31 }
  }

Will return 404 if the given C<id> does not match any records in the database.

=item * PUT /:source

Used to INSERT or UPDATE a single row. The HTTP body must be a JSON
structure with the data to insert or update.

The return value will be a JSON document containing all the data for the
inserted or updated row. Example:

  {
    "inserted": true, # or false
    "data": { "id": 1002, "name": "Jan Henning Thorsen", "age": 31 }
  }

=back

=head2 Error handling

The API will return "200 OK" on success and another error code on failure:

=over 4

=item * 400

Return the document below on invalid input data. C<message> holds a
description of what is missing. Example:

  { "message": "Missing (q) query param." }

=item * 401

Return the document below on when not authorized. C<message> holds a
description of why not. Example:

  { "message": "Invalid token." }

=item * 404

Return the document below if the given resource could not be found.
C<message> holds a description of what is not found. Examples:

  { "message": "Resource not found" }
  { "message": "No source by that name." }
  { "message": "No matching records in database." }

=item * 500

  { "message": "Internal server error." }

Generic error when something awful happens. C<message> might not make any
sense. Look at the server log for more details.

=back

Other error codes might be added in future releases.

=head2 Queries

The queries (referred to as the "q" query param in the API) are passed on as
the first argument to L<DBIx::Class/search>.

=cut

use Mojo::Base 'Mojolicious';
use File::Spec ();

our $VERSION = '0.05';

=head1 ATTRIBUTES

=head2 inactive_timeout

  $int = $self->inactive_timeout;

Used to set the number of seconds before a query agains the database time out.
Defaults to value from config, the environment variable
C<COOKIEVILLE_INACTIVE_TIMEOUT> or 10 seconds.

=head2 connect_args

  $array_ref = $self->connect_args;

Looks in L<config|Mojo/config> to find connection arguments for
L</schema_class>. See L</SYNOPSIS> for details.

=head2 schema_class

  $class_name = $self->schema_class;

Looks in L<config|Mojo/config> to find the schema class to use.
See L</SYNOPSIS> for details.

=cut

has inactive_timeout => sub {
  $ENV{COOKIEVILLE_INACTIVE_TIMEOUT} || shift->config('inactive_timeout') || 10;
};

has connect_args => sub {
  shift->config('connect_args') || [];
};

has schema_class => sub {
  shift->config('schema_class') || '';
};

=head1 HELPERS

=head2 db

  $obj = $self->db;

Returns an instance of L</schema_class>.

=head1 METHODS

=head2 setup_routes

Used to setup the L</RESOURCES>.

=cut

sub setup_routes {
  my $self = shift;
  my $r = $self->routes;

  if (my $rules = $self->config('access_rules')) {
    $self->plugin('Cookieville::Plugin::Authorize', $rules);
    $r = $r->find('cookieville_authorizer');
  }

  $self->routes->get('/')->to('schema#index')->name('cookieville') unless $r->find('cookieville');
  $r->get('/sources')->to('schema#sources_list')->name('schema_source_list') unless $r->find('schema_source_list');
  $r->get('/:source/schema')->to('schema#source_schema')->name('schema_for_source') unless $r->find('schema_for_source');
  $r->get('/:source/search')->to('read#search')->name('source_search') unless $r->find('source_search');
  $r->delete('/:source/:id')->to('write#delete')->name('source_delete') unless $r->find('source_delete');
  $r->patch('/:source/:id')->to('write#patch')->name('source_patch') unless $r->find('source_patch');
  $r->put('/:source')->to('write#update_or_insert')->name('source_update_or_insert') unless $r->find('source_update_or_insert');
}

=head2 startup

Will set up L</RESOURCES> and add L</HELPERS>.

=cut

sub startup {
  my $self = shift;

  if ($ENV{MOJO_CONFIG}) {
    $self->plugin('config');
  }
  if (my $config = $self->config('log')) {
    $config->{$_} and $self->log->$_($config->{$_}) for qw( level path );
    delete $self->log->{handle};
  }
  if (my $config = $self->config('access_log')) {
    $self->plugin('Cookieville::Plugin::AccessLog', $config);
  }
  if (my $schema_class = $self->schema_class) {
    eval "require $schema_class;1" or die $@;
  }

  $self->hook(before_dispatch => sub {
    my $c = shift;
    Mojo::IOLoop->stream($c->tx->connection)->timeout($self->inactive_timeout);
  });

  push @{ $self->renderer->classes }, __PACKAGE__;
  $self->defaults(format => 'json', message => '');
  $self->types->type(csv => 'text/csv');
  $self->helper(db => sub { shift->stash->{db} ||= $self->schema_class->connect(@{ $self->connect_args }); });
  $self->helper(j => sub { Mojo::JSON->new->encode($_[1]); });
  $self->setup_routes;
}

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014, Jan Henning Thorsen

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

=head1 AUTHOR

Jan Henning Thorsen - C<jhthorsen@cpan.org>

=cut

1;

__DATA__
@@ not_authorized.json.ep
%== j { message => $message || 'Not authorized.' }
@@ not_authorized.csv.ep
%= $message || 'Not authorized.'
@@ not_authorized.txt.ep
%= $message || 'Not authorized.'
@@ not_found.csv.ep
%= $message || 'Not found.'
@@ not_found.json.ep
%== j { message => $message || 'Not found.' }
@@ not_found.txt.ep
%= $message || 'Not found.'
@@ exception.csv.ep
%= $message || 'Internal server error.'
@@ exception.json.ep
%== j { message => $message || 'Internal server error.' }
@@ exception.txt.ep
%= $message || 'Internal server error.'
